posts/2024-09-05-impl-trait-capture-rules.md

```diff
@@ -0,0 +1,211 @@
+
+### incorrect suggestion
+
+Adding a `+ '_` argument to `impl Trait` may be confusing, but it's not terribly difficult. Unfortunately, it's often the wrong annotation, leading to unnecessary compiler errors -- and the *right* fix is either complex or sometimes not even possible. Consider an example like this:
+
+```rust
+fn process<T> {
+    context: &Context,
+    datums: Vec<T>,
+) -> impl Iterator<Item = ()> + '_ {
+    datums
+        .into_iter()
+        .map(|datum| context.process(datum))
+}
+```
+
+Here the `process` function applies `context.process` to each of the elements in `datums` (of type `T`). Because it captures `context`, it is declared as `+ '_`. This tells the compiler that the hidden type must outlive `'_`. Because this hidden type includes the `Vec<T>`, this in turn means that `T` must outlive `'_` -- but this is not true. So you get a compilation error ([try it on the playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b742fbf9b083a6e837db0b170489f34a)). 
+
+Just as before, this error is obscure, touching on the more complex aspects of Rust's type system. Unlike before, there is no easy fix! This problem in fact occurred frequently in the compiler, leading to an [obscure workaround called the `Captures` trait](https://github.com/rust-lang/rust/issues/34511#issuecomment-373423999). Gross!
+
+We surveyed crates on crates.io and found that the vast majority of cases involving return-position impl trait and generics had bounds that were too strong and which could lead to unnecessary errors (though often they were used in simple ways that didn't trigger an error).
+
+### inconsistencies with other parts of Rust
+
+The current design was also introducing inconsistencies with other parts of Rust.
+
+#### async fn desugaring
+
+Rust defines an `async fn` as desugaring to a normal `fn` that returns `-> impl Future`. You might therefore expect that a function like `process`:
+
+```rust
+async fn process(data: &Data) { .. }
+```
+
+...would be (roughly) desugared to:
+
+```rust
+fn process(
+    data: &Data
+) -> impl Future<Output = ()> {
+    async move {
+        ..
+    }
+}
+```
+
+In practice, because of the problems with lifetime capture, this is not the actual desugaring. The actual desugaring is to a special kind of `impl Trait` that is allowed to capture all lifetimes. But that form of `impl Trait` was not exposed to end-users.
+
+#### impl trait in traits
+
+As we pursued the design for impl trait in traits ([RFC 3425](https://rust-lang.github.io/rfcs/3425-return-position-impl-trait-in-traits.html)), we encountered a number of challenges related to the capturing of lifetimes. [In order to get the symmetries that we wanted to work](https://hackmd.io/zgairrYRSACgTeZHP1x0Zg) (e.g., that one can write `-> impl Future` in a trait and impl with the expected effect), we had to change the capture rules to capture all lifetime parameters uniformly.
```

Comment:
It might be worth reiterating this later on in the post: I doubt most users are aware that RPITIT capture rules work differently.